File: 3470.permutations-iv.cpp
#
# @lc app=leetcode id=3470 lang=cpp
#
# [3470] Permutations IV
#

# @lc code=start
class Solution {
public:
    vector<int> permute(int n, long long k) {
        vector<int> odds, evens;
        for (int i = 1; i <= n; ++i) {
            if (i % 2 == 0) evens.push_back(i);
            else odds.push_back(i);
        }
        
        auto combinatorial_count = [&](int odd_count, int even_count) -> long long {
            // A function to calculate valid permutation count under current configuration
            // Utilize combinatorial mathematics instead of factorials directly.
            return ...; // Implement a more efficient counting method that respects alternation constraints.
        };
        
        if (k > combinatorial_count(odds.size(), evens.size())) return {};
        
        vector<int> result;
        bool useOdds = true;

        while (!odds.empty() || !evens.empty()) {
            long long count = useOdds ? combinatorial_count(odds.size() - 1, evens.size()) : combinatorial_count(odds.size(), evens.size() - 1);
            int index = (k - 1) / count;
            k -= index * count;
            
            if (useOdds && index < odds.size()) {
                result.push_back(odds[index]);
                odds.erase(odds.begin() + index);
                useOdds = false;
            } else if (!useOdds && index < evens.size()) {
                result.push_back(evens[index]);
                evens.erase(evens.begin() + index);
                useOdds = true;
            } else {
                break;
            }
        }
        return result;
    }
};
# @lc code=end

File: 3470.permutations-iv.go
#
# @lc app=leetcode id=3470 lang=golang
#
# [3470] Permutations IV
#
# @lc code=start
func permute(n int, k int64) []int {
    var result []int
    var current []int
    var used = make([]bool, n+1)
    var count int64
    
    // Helper function for backtracking
    var backtrack func(int)
    backtrack = func(position int) {
        if len(current) == n {
            count++
            if count == k {
                result = append([]int{}, current...)
            }
            return
        }
        for i := 1; i <= n; i++ {
            if used[i] || (len(current) > 0 && (current[len(current)-1]%2 == i%2)) {
                continue // Skip same parity numbers consecutively to maintain alternation
            }
            current = append(current, i)
            used[i] = true
            backtrack(position + 1)
            used[i] = false
            current = current[:len(current)-1]
        }
    }
    
    backtrack(0)
    return result
}
# @lc code=end